### ISSUE TITLE
retraining DeepVariant

### ISSUE THREAD

**Issue by sophienguyen01** [google/deepvariant#698]
I'm interested in retraining DeepVariant to work on our dataset with low allele fraction variants. In this [tutorial](https://github.com/google/deepvariant/blob/ab068c4588a02e2167051bd9e74c0c9579462b51/docs/deepvariant-training-case-study.md), the training and validation set only target one chromosome region, chromosome 1 and 21 respectively.  I want to create training samples on all chromosome regions in .BAM files with parameters to select for low fraction variants (using --downsample_fraction and vsc_min_fraction_snps). Is it a good way to do or do you suggest I break the training examples into different chromosome regions? 
Also if I use one BAM file and create training set on all chr regions of that file, should I use another BAM file to create the validation set?

Thank you

**Comment by pgrosu** [comment]
Hi Sophie,

Ambitious endeavor, but some things to be aware of.  So a low VAF would probably confuse the model training, as that is usually a homozygous reference call.  With a low VAF, the other alleles would probably be automatically be selected as supporting information for training in the pileup to confirm a call, as they would be in the majority.  Thus when you run your trained model in the future against new BAM files, it would not look so much for the rare variants, but rather select for the more prevalent ones in your data to call a genotype (GT) and subsequently a GQ score of supporting it.  

Regarding training and validation, they always usually must be different.  For example, in the original DeepVariant paper under the [Supplementary Text and Figures](https://static-content.springer.com/esm/art%3A10.1038%2Fnbt.4235/MediaObjects/41587_2018_BFnbt4235_MOESM81_ESM.pdf), they trained on chromosomes 1-19, and validated on 20-22:

![image](https://github.com/google/deepvariant/assets/6555937/83cdb7f3-44e9-4185-99cf-81362489dfcf)

If you want to train chromosome-by-chromosome, it is possible to take a previously trained model and continue training it as noted in the [tutorial specified by `GCS_PRETRAINED_WGS_MODEL`](https://github.com/google/deepvariant/blob/ab068c4588a02e2167051bd9e74c0c9579462b51/docs/deepvariant-training-case-study.md#start-model_train-and-model_eval), but that could create batch effects (meaning it will shift the model one way with one chromosome, and then back with another even under shuffling conditions).  So it might be better to start with a complete set of chromosomes you select for training.

Not trying to discourage you, so give it a try and let's see what happens -- in any case it will be fun learning experience.

Hope it helps,
Paul

**Comment by MariaNattestad** [comment]
@sophienguyen01 For DeepVariant production models we generally train on chr1-19, tune on chr21-22, and save chr20 for final "test" or "inference" evaluations. When we have enough samples we'll leave out a whole sample; for most (maybe all) of our production models this is currently HG003 that is never seen during training (or tune), only kept for inference.

It's important to do this train/tune/test split at the sample level and/or chromosome level so DeepVariant can't overfit to the specific variants it sees, which it could if you for example left out one of two bam files that came from the same sample.

**Comment by akolesnikov** [comment]
Closing the issue for now. Feel free to reopen if needed.

**Comment by sophienguyen01** [comment]
HI,

I want to ask about shuffling. @akolesnikov has a good explanation on shuffling on this issue (#312). My training has multiple  samples (HG001 to HG007 except HG003), so I don't know if shuffling means shuffling within samples or within all samples. Thank you

**Comment by pgrosu** [comment]
Hi Sophie,

You have to train across all the samples.  The reason is that the model will need to be representative of all those samples, and you don't want to overfit for one sample over another.  Today there is more data than computing power and memory, so training happens by batching the data, training on a batch, and then tweaking the model in the follow-up batches.  Here's a visual example of what could happen:

![image](https://github.com/google/deepvariant/assets/6555937/3ff13990-c2cc-4863-a904-bb0219791b06)

In the first batch, the coefficients will be biased in a monotonically increasing function.  Then when the second batch is used to tweak the model's parameters, it will shift towards a monotonically decreasing function shifting it towards an opposite direction. Then when the third batch comes in it will shift the parameters in the other direction.  Sure, the independent variable (x-axis) here is a simple one-dimensional type, but this can become complex with multiple channels (`read base`, `base quality`, `mapping quality`, etc).

As @akolesnikov mentioned, the `input_pattern_list` parameter of the shuffle script takes a list of files, so you can do it all at once.

Let me know if I should expand on anything.

Hope it helps,
Paul

**Comment by sophienguyen01** [comment]
Thank you, that helps a lot.

**Comment by pgrosu** [comment]
You're welcome Sophie.
